corehq/apps/hqwebapp/signals.py

```diff
@@ -26,9 +27,27 @@ def clear_failed_logins_and_unlock_account(sender, request, user, **kwargs):
 
 
 @receiver(user_login_failed)
-def add_failed_attempt(sender, credentials, **kwargs):
+def add_failed_attempt(sender, credentials, token_failure=False, **kwargs):
     user = CouchUser.get_by_username(credentials['username'])
-    if user and not user.is_locked_out() and user.supports_lockout():
+    if not user:
+        metrics_counter('commcare.auth.invalid_user')
+        return
+
+    if token_failure:
+        metrics_counter('commcare.auth.invalid_token')
+
+    locked_out = user.is_locked_out()
+
+    if locked_out:
+        lockout_result = 'locked_out'
+    else:
+        lockout_result = 'should_be_locked_out' if user.should_be_locked_out() else 'allowed_to_retry'
+
+    metrics_counter('commcare.auth.failed_attempts', tags={
+        'result': lockout_result
+    })
+
+    if not locked_out:
```

Comment:
This used to only apply if `user.supports_lockout()` whereas now it applies regardless which has caused some ResourceConflicts: https://sentry.io/organizations/dimagi/issues/2186196976/?project=136860&query=is%3Aunresolved

One option could be to change it to: `if not locked_out and user.supports_lockout():` or else we if we want to always track login attempts we could handle the ResourceConflict and re-fetch the doc.

Reply:
I've [reverted](https://github.com/dimagi/commcare-hq/pull/29063) this PR for now and re-opened it here: https://github.com/dimagi/commcare-hq/pull/29064

Reply:
Yeah, this was intentional. We were concerned that `supports_lockout()` always returned False for mobile users. Given that mobile also contains sensitive data, we wanted a metric to understand how big of a problem this was

Reply:
yea, I think for now we could do `if not locked_out and user.supports_lockout():`

Reply:
@dannyroberts thoughts? It means we won't be able to track mobile attempts

Reply:
I think it's totally reasonable to change the "redo" of this PR to not support reporting would-be lockouts for mobile yet by changing it back to `if not locked_out and user.supports_lockout()`, then in a followup PR add that back.

As for how you add it back, what you said about storing the failed login attempts in another model makes sense and is a valid approach, but we might be able to do something simpler by understanding how this `ResourceConflict` happens.

A couch ResourceConflict happens when you save a doc whose '_rev' field doesn't match the _rev value currently in the db. _rev automatically changes every time a doc is saved, so it basically acts as a way for you to prove to CouchDB, when you save the doc, that you have the latest version of the doc. In practice a `ResourceConflict` mean (1) you fetched a stale copy of the doc somehow and then saved it (2) you fetched the doc twice in a row, saved the first one, and then tried to save the second one (the first save changes the rev in the db and for the first copy but not for the second copy) (3) you and another process fetched the doc at the same time, the other process saved it, and then you saved it (this is basically the same as #2, but with a race condition rather than it all being sequential in a single thread).

I think what's happening here is actually (1). (3) is possible but unlikely given how often it's happening, and (2) is unlikely because it looks like you're fetching it at the top of the function and this is the only save in the function and the sentry error shows it error. (By here I mean 6 lines below.) So I think what's happening is that `CouchUser.get_by_username(credentials['username'])` is returning a stale copy of the user doc. If you use `CouchUser.get_by_username(credentials['username'], strict=True)` instead, I think that might solve the issue. You chan check out that function, but it has a arg strict that defaults to False that dictates whether it's okay with a slightly stale copy of the doc in exchange for speed (fine when you're just looking up the email or something that doesn't change much) (strict=False) or you need the actual latest copy of the doc (strict=True). When you're planning to save it, `strict=True` is important.

Reply:
Thanks for the detailed writeup, @dannyroberts. One concern I have, though, is I think we're still talking about a pseudo-race condition strategy, here. For there to be a resource conflict, we're proposing that we're fetching a stale document from our cache. However, [we clear the cache on every save](https://github.com/dimagi/commcare-hq/blob/300c80cb77c097cb3924f6601c5d7d9baf19902e/corehq/apps/users/models.py#L1535), so as far as I can tell, the only way for us to fetch a stale document is for multiple nodes to be trying to save the same user within a 5 minute window (it looks like quickcache expires after 5 minutes). Given the frequency that this conflict happened (it went from ~0 to 163 in one day), I'm not sure if that theory would explain it. That said, I think the change you're proposing is pretty safe -- it shouldn't be a big deal to force a re-fetch on users when they log in, right?

Reply:
Yeah I mean try and and see if it fixes the similar test failure you had worked around. Yeah it shouldn't be a big deal to re-fetch the user doc when a user logs in

Reply:
Reverted again: https://github.com/dimagi/commcare-hq/pull/29097